### PR TITLE
Nestedpayload

### PR DESCRIPTION
**Pull request by jdonszelmann**
adds NestedPacket struct in conformance with the NestedPayload class in py-ipv8

### CODE REVIEW

File: src/networking/payloads/binmemberauthenticationpayload.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+use serde::{Serialize,Deserialize};
+use crate::networking::payloads::Ipv8Payload;
+use crate::networking::serialization::varlen::VarLen16;
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct BinMemberAuthenticationPayload {
+  /// The global time in the system. Uses a lamport clock system.
```
Comment (NULLx76) [inline src/networking/payloads/binmemberauthenticationpayload.rs]: This comment seems out of place

File: src/networking/payloads/mod.rs [hunk 2]
```diff
@@ -1,6 +1,12 @@
 // module attempting to ensure feature parity with py-ipv8/ipv8/messaging/payload.py
 pub mod connectiontype;
+
```
Comment (NULLx76) [inline src/networking/payloads/mod.rs]: No need for a newline here

File: src/networking/serialization/mod.rs [hunk 3]
```diff
@@ -97,6 +120,28 @@ mod tests {
     // doesnt have anything but needed for the default implementation (as of right now)
   }
 
+// only works with feature(test) and with `extern crate test; use test::Bencher`
```
Comment (NULLx76) [inline src/networking/serialization/mod.rs]: Should probably be removed and replaced in favor of proper benchmarks with `Criterion.rs`

File: src/networking/serialization/nestedpayload.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    let i = VarLen16((self.0).0.to_owned());
+
+    let mut state = serializer.serialize_struct("NestedPacket", self.0.len())?;
+    state.serialize_field("payload", &i)?;
+    state.end()
+  }
+}
+
+#[derive(Debug, PartialEq, serde::Deserialize)]
+/// this is the actual pattern of an NestedPayload.z
```
Comment (NULLx76) [inline src/networking/serialization/nestedpayload.rs]: Dangling `z` and minor spelling mistake `an` should be `a`.

File: src/networking/serialization/nestedpayload.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  #[derive(Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+  struct TestPayload4 {
+    test: Vec<u8>
+  }
+
+  impl Ipv8Payload for TestPayload4 {
+    // doesnt have anything but needed for the default implementation (as of right now)
+  }
+
+
```
Comment (NULLx76) [inline src/networking/serialization/nestedpayload.rs]: Unnecessary newline

File: tests/testpyipv8packets.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      },
+      source_wan_address: Address {
+        address: Ipv4Addr::new(192,168,1,75),
+        port: 8090
+      },
+      // TODO i am not entirely sure of the extraction of bits is 100% right. though public does seem plausible.
+      advice: false,
+      connection_type: ConnectionType::PUBLIC,
+      identifier:151,
+      // TODO: i have not at all verified if this RawEnd is correct yet.
```
Comment (NULLx76) [inline tests/testpyipv8packets.rs]: Do we need to resolve these todos before merging?

File: tests/testpyipv8packets.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      },
+      source_lan_address: Address {
+        address: Ipv4Addr::new(192,168,1,75),
+        port: 8090
+      },
+      source_wan_address: Address {
+        address: Ipv4Addr::new(192,168,1,75),
+        port: 8090
+      },
+      // TODO i am not entirely sure of the extraction of bits is 100% right. though public does seem plausible.
```
Comment (NULLx76) [inline tests/testpyipv8packets.rs]: Do we need to resolve these todos before merging?

**Review by NULLx76 — CHANGES_REQUESTED** [review]

File: src/networking/serialization/mod.rs (hunk 3 above)
Comment (NULLx76) [reply src/networking/serialization/mod.rs]: Doesn't necessarily have to be done right now though

Comment (jdonszelmann) [reply src/networking/serialization/mod.rs]: not doing right now. making an issue in the project board

File: tests/testpyipv8packets.rs (hunk 6 above)
Comment (jdonszelmann) [reply tests/testpyipv8packets.rs]: not really. lets support headers and signatures and then test on full packets :D

Comment (jdonszelmann) [reply tests/testpyipv8packets.rs]: thats easier too. these tests were mostly to impress Tribler

File: tests/testpyipv8packets.rs (hunk 7 above)
Comment (jdonszelmann) [reply tests/testpyipv8packets.rs]: same

**Review by NULLx76 — APPROVED** [review]

**Review by dsluijk — APPROVED** [review]

**Pull request merged**